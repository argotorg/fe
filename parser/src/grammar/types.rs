use crate::ast::{ContractStmt, EventField, Field, GenericArg, ModuleStmt, TypeDesc};
use crate::grammar::expressions::parse_expr;
use crate::node::{Node, Span};
use crate::{ParseFailed, ParseResult, Parser, TokenKind};

/// Parse a [`ModuleStmt::StructDef`].
/// # Panics
/// Panics if the next token isn't `struct`.
pub fn parse_struct_def(par: &mut Parser) -> ParseResult<Node<ModuleStmt>> {
    use TokenKind::*;

    let struct_tok = par.assert(Struct);
    let name = par.expect_with_notes(Name, "failed to parse struct definition", || {
        vec!["Note: a struct name must start with a letter or underscore, and contain letters, numbers, or underscores".into()]
    })?;

    let mut fields = vec![];
    par.enter_block(struct_tok.span + name.span, "struct definition")?;
    loop {
        match par.peek() {
            Some(Name) | Some(Pub) | Some(Const) => {
                let pub_qual = parse_opt_qualifier(par, TokenKind::Pub);
                let const_qual = parse_opt_qualifier(par, TokenKind::Const);
                fields.push(parse_field(par, pub_qual, const_qual)?);
            }
            Some(Dedent) => {
                par.next()?;
                break;
            }
            None => break,
            Some(_) => {
                let tok = par.next()?;
                par.unexpected_token_error(tok.span, "failed to parse struct def", vec![]);
                return Err(ParseFailed);
            }
        }
    }
    let span = struct_tok.span + name.span + fields.last();
    Ok(Node::new(
        ModuleStmt::StructDef {
            name: name.into(),
            fields,
        },
        span,
    ))
}

/// Parse a type definition, e.g. `type MyMap = map<u8, address>`.
/// # Panics
/// Panics if the next token isn't `type`.
pub fn parse_type_def(par: &mut Parser) -> ParseResult<Node<ModuleStmt>> {
    let type_tok = par.assert(TokenKind::Type);
    let name = par.expect(TokenKind::Name, "failed to parse type declaration")?;
    par.expect_with_notes(TokenKind::Eq, "failed to parse type declaration", || {
        vec![
            "Note: a type alias name must be followed by an equals sign and a type description"
                .into(),
            format!("Example: `type {} = map<address, u64>`", name.text),
        ]
    })?;
    let typ = parse_type_desc(par)?;
    let span = type_tok.span + typ.span;
    Ok(Node::new(
        ModuleStmt::TypeDef {
            name: name.into(),
            typ,
        },
        span,
    ))
}

/// Parse an event definition.
/// # Panics
/// Panics if the next token isn't `event`.
pub fn parse_event_def(par: &mut Parser) -> ParseResult<Node<ContractStmt>> {
    use TokenKind::*;

    let event_tok = par.assert(Event);
    let name = par.expect(Name, "failed to parse event definition")?;

    let mut fields = vec![];
    par.enter_block(event_tok.span + name.span, "event definition")?;
    loop {
        match par.peek() {
            Some(Name) | Some(Idx) => {
                fields.push(parse_event_field(par)?);
            }
            Some(Dedent) => {
                par.next()?;
                break;
            }
            None => break,
            Some(_) => {
                let tok = par.next()?;
                par.unexpected_token_error(tok.span, "failed to parse event definition", vec![]);
                return Err(ParseFailed);
            }
        }
    }
    let span = event_tok.span + name.span + fields.last();
    Ok(Node::new(
        ContractStmt::EventDef {
            name: name.into(),
            fields,
        },
        span,
    ))
}

/// Parse an event field, e.g. `foo: u8` or `idx from: address`.
pub fn parse_event_field(par: &mut Parser) -> ParseResult<Node<EventField>> {
    let idx_qual = parse_opt_qualifier(par, TokenKind::Idx);
    let name = par.expect(TokenKind::Name, "failed to parse event field")?;
    par.expect_with_notes(TokenKind::Colon, "failed to parse event field", || {
        vec![
            "Note: event field name must be followed by a colon and a type description".into(),
            format!(
                "Example: `{}{}: address`",
                if idx_qual.is_some() { "idx " } else { "" },
                name.text
            ),
        ]
    })?;

    let typ = parse_type_desc(par)?;
    par.expect_newline("event field")?;
    let span = name.span + idx_qual + &typ;
    Ok(Node::new(
        EventField {
            is_idx: idx_qual.is_some(),
            name: Node::new(name.text.into(), name.span),
            typ,
        },
        span,
    ))
}

/// Parse a field for a struct or contract. The leading optional `pub` and
/// `const` qualifiers must be parsed by the caller, and passed in.
/// Note that `event` fields are handled in [`parse_event_field`].
pub fn parse_field(
    par: &mut Parser,
    pub_qual: Option<Span>,
    const_qual: Option<Span>,
) -> ParseResult<Node<Field>> {
    let name = par.expect(TokenKind::Name, "failed to parse field definition")?;
    par.expect_with_notes(TokenKind::Colon, "failed to parse field definition", || {
        vec![
            "Note: field name must be followed by a colon and a type description".into(),
            format!(
                "Example: {}{}{}: address",
                if pub_qual.is_some() { "pub " } else { "" },
                if const_qual.is_some() { "const " } else { "" },
                name.text
            ),
        ]
    })?;

    let typ = parse_type_desc(par)?;
    let value = if par.peek() == Some(TokenKind::Eq) {
        par.next()?;
        Some(parse_expr(par)?)
    } else {
        None
    };
    par.expect_newline("field definition")?;
    let span = name.span + pub_qual + const_qual + &typ;
    Ok(Node::new(
        Field {
            is_pub: pub_qual.is_some(),
            is_const: const_qual.is_some(),
            name: Node::new(name.text.into(), name.span),
            typ,
            value,
        },
        span,
    ))
}

/// Parse an optional qualifier (`pub`, `const`, or `idx`).
pub fn parse_opt_qualifier(par: &mut Parser, tk: TokenKind) -> Option<Span> {
    if par.peek() == Some(tk) {
        let tok = par.next().unwrap();
        Some(tok.span)
    } else {
        None
    }
}

/// Parse an angle-bracket-wrapped list of generic arguments (eg. the tail end
/// of `map<address, u256>`).
/// # Panics
/// Panics if the first token isn't `<`.
fn parse_generic_args(par: &mut Parser) -> ParseResult<(Vec<Node<GenericArg>>, Span)> {
    use TokenKind::*;
    let mut span = par.assert(Lt).span;

    let mut args = vec![];

    loop {
        match par.peek_or_err()? {
            Gt => {
                span += par.next()?.span;
                break;
            }
            GtGt => {
                span += par.split_next()?.span;
                break;
            }
            Int => {
                let tok = par.next()?;
                if let Ok(num) = tok.text.parse() {
                    args.push(Node::new(GenericArg::Int(num), tok.span));
                } else {
                    par.error(tok.span, "failed to parse integer literal");
                    return Err(ParseFailed);
                }
            }
            Name | ParenOpen => {
                let typ = parse_type_desc(par)?;
                args.push(Node::new(GenericArg::TypeDesc(typ.kind), typ.span));
                if par.peek() == Some(Comma) {
                    par.next()?;
                } else {
                    // If there's no comma, the next token must be `>`
                    match par.peek_or_err()? {
                        Gt => {
                            span += par.next()?.span;
                            break;
                        }
                        GtGt => {
                            span += par.split_next()?.span;
                            break;
                        }
                        _ => {
                            let tok = par.next()?;
                            par.unexpected_token_error(
                                tok.span,
                                "Unexpected token while parsing generic arg list",
                                vec![],
                            );
                            return Err(ParseFailed);
                        }
                    }
                }
            }
            _ => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    tok.span,
                    "failed to parse generic type argument list",
                    vec![],
                );
                return Err(ParseFailed);
            }
        }
    }
    Ok((args, span))
}

/// Parse a type description, e.g. `u8` or `map<address, u256>`.
pub fn parse_type_desc(par: &mut Parser) -> ParseResult<Node<TypeDesc>> {
    use TokenKind::*;

    let mut typ = match par.peek_or_err()? {
        Name => {
            let name = par.next()?;
            match par.peek() {
                Some(Lt) => {
                    let (args, argspan) = parse_generic_args(par)?;
                    let span = name.span + argspan;
                    Node::new(
                        TypeDesc::Generic {
                            base: name.into(),
                            args,
                        },
                        span,
                    )
                }
                _ => Node::new(
                    TypeDesc::Base {
                        base: name.text.into(),
                    },
                    name.span,
                ),
            }
        }
        ParenOpen => {
            let mut span = par.next()?.span;
            let mut items = vec![];
            loop {
                match par.peek_or_err()? {
                    ParenClose => {
                        span += par.next()?.span;
                        break;
                    }

                    Name | ParenOpen => {
                        let item = parse_type_desc(par)?;
                        span += item.span;
                        items.push(item);
                        if par.peek_or_err()? == Comma {
                            par.next()?;
                        } else {
                            span += par
                                .expect(
                                    ParenClose,
                                    "Unexpected token while parsing tuple type description",
                                )?
                                .span;
                            break;
                        }
                    }

                    _ => {
                        let tok = par.next()?;
                        par.unexpected_token_error(
                            tok.span,
                            "failed to parse type description",
                            vec![],
                        );
                        return Err(ParseFailed);
                    }
                }
            }
            Node::new(TypeDesc::Tuple { items }, span)
        }
        _ => {
            let tok = par.next()?;
            par.unexpected_token_error(tok.span, "failed to parse type description", vec![]);
            return Err(ParseFailed);
        }
    };

    while par.peek() == Some(BracketOpen) {
        let ctx = "Unexpected token while parsing array type description.";
        let mut span = typ.span + par.next()?.span;
        let num = par.expect(Int, ctx)?;
        if let Ok(dimension) = num.text.parse() {
            span += par.expect(BracketClose, ctx)?.span;
            typ = Node::new(
                TypeDesc::Array {
                    typ: Box::new(typ),
                    dimension,
                },
                span,
            );
        } else {
            par.error(num.span, "failed to parse number literal");
            return Err(ParseFailed);
        }
    }
    Ok(typ)
}
