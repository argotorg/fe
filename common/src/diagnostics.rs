use crate::files::{FileStore, SourceFileId};
pub use codespan_reporting::diagnostic::{Diagnostic as CsDiagnostic, Label, LabelStyle, Severity};
use codespan_reporting::term;
use term::termcolor::{BufferWriter, ColorChoice};

pub type Diagnostic = CsDiagnostic<SourceFileId>;

/// Print the given diagnostics to stderr.
pub fn print_diagnostics(diagnostics: &[Diagnostic], files: &FileStore) {
    let writer = BufferWriter::stderr(ColorChoice::Auto);
    let mut buffer = writer.buffer();
    let config = term::Config::default();

    for diag in diagnostics {
        term::emit(&mut buffer, &config, files, &diag).unwrap();
    }
    // If we use `writer` here, the output won't be captured by rust's test system.
    eprintln!("{}", std::str::from_utf8(buffer.as_slice()).unwrap());
}

/// Format the given diagnostics as a string.
pub fn diagnostics_string(diagnostics: &[Diagnostic], files: &FileStore) -> String {
    let writer = BufferWriter::stderr(ColorChoice::Never);
    let mut buffer = writer.buffer();
    let config = term::Config::default();

    for diag in diagnostics {
        term::emit(&mut buffer, &config, files, &diag).expect("failed to emit diagnostic");
    }
    std::str::from_utf8(buffer.as_slice()).unwrap().to_string()
}
