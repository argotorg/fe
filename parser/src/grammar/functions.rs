use super::expressions::{parse_call_args, parse_expr};
use super::types::parse_type_desc;

use crate::ast::{BinOperator, ContractStmt, Expr, FuncDefArg, FuncStmt, VarDeclTarget};
use crate::lexer::TokenKind;
use crate::node::{Node, Span};
use crate::{Label, ParseFailed, ParseResult, Parser};

/// Parse a function definition. The optional `pub` qualifier must be parsed by
/// the caller, and passed in.
///
/// # Panics
/// Panics if the next token isn't `def`.
pub fn parse_fn_def(par: &mut Parser, pub_qual: Option<Span>) -> ParseResult<Node<ContractStmt>> {
    let def_tok = par.assert(TokenKind::Def);
    let name = par.expect(TokenKind::Name, "failed to parse function definition")?;
    let mut span = def_tok.span + pub_qual + name.span;

    let args = match par.peek_or_err()? {
        TokenKind::ParenOpen => {
            let node = parse_fn_param_list(par)?;
            span += node.span;
            node.kind
        }
        TokenKind::Colon | TokenKind::Arrow => {
            par.fancy_error(
                "function definition requires a list of parameters",
                vec![Label::primary(
                    name.span,
                    "function name must be followed by `(`".into(),
                )],
                vec![
                    format!(
                        "Note: if the function `{}` takes no parameters, use an empty set of parentheses.",
                        name.text
                    ),
                    format!("Example: def {}()", name.text),
                    "Note: each parameter must have a name and a type.".into(),
                    format!("Example: def {}(my_value: u256, x: bool)", name.text),
                ],
            );
            vec![]
        }
        _ => {
            let tok = par.next()?;
            par.unexpected_token_error(
                tok.span,
                "failed to parse function definition",
                vec![
                    "function name must be followed by a list of parameters".into(),
                    "Example: `def foo(x: address, y: u256)` or `def f()`".into(),
                ],
            );
            return Err(ParseFailed);
        }
    };
    let return_type = if par.peek() == Some(TokenKind::Arrow) {
        par.next()?;
        Some(parse_type_desc(par)?)
    } else {
        None
    };
    span += return_type.as_ref();

    // TODO: allow multi-line return type? `def f()\n ->\n u8`
    // TODO: allow single-line fn defs?
    par.enter_block(span, "function definition")?;
    let body = parse_block_stmts(par)?;
    span += body.last();
    Ok(Node::new(
        ContractStmt::FuncDef {
            is_pub: pub_qual.is_some(),
            name: name.into(),
            args,
            return_type,
            body,
        },
        span,
    ))
}

fn parse_fn_param_list(par: &mut Parser) -> ParseResult<Node<Vec<Node<FuncDefArg>>>> {
    let mut span = par.assert(TokenKind::ParenOpen).span;
    let mut params = vec![];
    loop {
        match par.peek_or_err()? {
            TokenKind::ParenClose => {
                span += par.next()?.span;
                break;
            }
            TokenKind::Name => {
                let name = par.next()?;

                par.expect_with_notes(
                    TokenKind::Colon,
                    "failed to parse function parameter",
                    || {
                        vec![
                            "Note: parameter name must be followed by a colon and a type description"
                                .into(),
                            format!("Example: `{}: u256`", name.text),
                        ]
                    },
                )?;
                let typ = parse_type_desc(par)?;
                let param_span = name.span + typ.span;
                params.push(Node::new(
                    FuncDefArg {
                        name: Node::new(name.text.into(), name.span),
                        typ,
                    },
                    param_span,
                ));

                if par.peek() == Some(TokenKind::Comma) {
                    par.next()?;
                } else {
                    span += par
                        .expect(
                            TokenKind::ParenClose,
                            "unexpected token while parsing function parameter list",
                        )?
                        .span;
                    break;
                }
            }
            _ => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    tok.span,
                    "failed to parse function parameter list",
                    vec![],
                );
                return Err(ParseFailed);
            }
        }
    }
    Ok(Node::new(params, span))
}

/// Parse (function) statements until a block dedent or end-of-file is reached.
fn parse_block_stmts(par: &mut Parser) -> ParseResult<Vec<Node<FuncStmt>>> {
    let mut body = vec![];
    loop {
        match par.peek() {
            None => break,
            Some(TokenKind::Dedent) => {
                par.next()?;
                break;
            }
            Some(_) => body.push(parse_stmt(par)?),
        }
    }
    Ok(body)
}

fn aug_assign_op(tk: TokenKind) -> Option<BinOperator> {
    use BinOperator::*;
    use TokenKind::*;

    let op = match tk {
        PlusEq => Add,
        MinusEq => Sub,
        StarEq => Mult,
        SlashEq => Div,
        PercentEq => Mod,
        StarStarEq => Pow,
        LtLtEq => LShift,
        GtGtEq => RShift,
        PipeEq => BitOr,
        HatEq => BitXor,
        AmperEq => BitAnd,
        _ => return None,
    };
    Some(op)
}

/// Parse a `continue`, `break`, `pass`, or `revert` statement.
///
/// # Panics
/// Panics if the next token isn't one of the above.
pub fn parse_single_word_stmt(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    let tok = par.next()?;
    par.expect_newline(tok.kind.symbol_str().unwrap())?;
    let stmt = match tok.kind {
        TokenKind::Continue => FuncStmt::Continue,
        TokenKind::Break => FuncStmt::Break,
        TokenKind::Pass => FuncStmt::Pass,
        TokenKind::Revert => FuncStmt::Revert,
        _ => panic!(),
    };
    Ok(Node::new(stmt, tok.span))
}

/// Parse a function-level statement.
pub fn parse_stmt(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    use TokenKind::*;

    // rule: stmt parsing fns eat the trailing separator (newline, semi, eof)
    match par.peek_or_err()? {
        For => parse_for_stmt(par),
        If => parse_if_stmt(par),
        While => parse_while_stmt(par),
        Return => parse_return_stmt(par),
        Assert => parse_assert_stmt(par),
        Continue | Break | Pass | Revert => parse_single_word_stmt(par),
        Emit => parse_emit_statement(par),
        _ => parse_expr_stmt(par),
    }
}

/// Parse a (function) statement that begins with an expression. This might be
/// a `VarDecl`, `Assign`, or an expression.
fn parse_expr_stmt(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    use TokenKind::*;
    let expr = parse_expr(par)?;
    let node = match par.peek() {
        None | Some(Newline) => {
            let span = expr.span;
            Node::new(FuncStmt::Expr { value: expr }, span)
        }
        Some(Colon) => {
            par.next()?;

            let target = expr_to_vardecl_target(par, expr)?;
            let typ = parse_type_desc(par)?;
            let value = if par.peek() == Some(Eq) {
                par.next()?;
                Some(parse_expr(par)?)
            } else {
                None
            };
            let span = target.span + typ.span + value.as_ref();
            Node::new(FuncStmt::VarDecl { target, typ, value }, span)
        }
        Some(Eq) => {
            par.next()?;
            let value = parse_expr(par)?;
            let span = expr.span + value.span;
            // TODO: should `x = y = z` be allowed?
            Node::new(
                FuncStmt::Assign {
                    target: expr,
                    value,
                },
                span,
            )
        }
        Some(tk) => {
            if let Some(op) = aug_assign_op(tk) {
                let op_tok = par.next()?;
                let value = parse_expr(par)?;
                let span = expr.span + value.span;
                Node::new(
                    FuncStmt::AugAssign {
                        target: expr,
                        op: Node::new(op, op_tok.span),
                        value,
                    },
                    span,
                )
            } else {
                let tok = par.next()?;
                par.unexpected_token_error(tok.span, "invalid syntax in function body", vec![]);
                return Err(ParseFailed);
            }
        }
    };
    par.expect_newline("statement")?;
    Ok(node)
}

fn expr_to_vardecl_target(par: &mut Parser, expr: Node<Expr>) -> ParseResult<Node<VarDeclTarget>> {
    match expr.kind {
        Expr::Name(name) => Ok(Node::new(VarDeclTarget::Name(name), expr.span)),
        Expr::Tuple { elts } if !elts.is_empty() => Ok(Node::new(
            VarDeclTarget::Tuple(
                elts.into_iter()
                    .map(|elt| expr_to_vardecl_target(par, elt))
                    .collect::<ParseResult<Vec<_>>>()?,
            ),
            expr.span,
        )),
        _ => {
            par.fancy_error(
                "failed to parse variable declaration",
                vec![Label::primary(
                    expr.span,
                    "invalid variable declaration target".into(),
                )],
                vec![
                    "The left side of a variable declaration can be either a name\nor a non-empty tuple."
                        .into(),
                ],
            );
            Err(ParseFailed)
        }
    }
}

/// Parse an `if` statement, or an `elif` block.
///
/// # Panics
/// Panics if the next token isn't `if` or `elif`.
pub fn parse_if_stmt(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    let if_tok = par.next()?;
    assert!(matches!(if_tok.kind, TokenKind::If | TokenKind::Elif));

    let test = parse_expr(par)?;
    par.enter_block(if_tok.span + test.span, "`if` statement")?;
    let body = parse_block_stmts(par)?;

    let else_block = match par.peek() {
        Some(TokenKind::Else) => {
            let else_tok = par.next()?;
            par.enter_block(else_tok.span, "`if` statement `else` branch")?;
            parse_block_stmts(par)?
        }
        Some(TokenKind::Elif) => {
            vec![parse_if_stmt(par)?]
        }
        _ => vec![],
    };

    let span = if_tok.span + test.span + body.last() + else_block.last();
    Ok(Node::new(
        FuncStmt::If {
            test,
            body,
            or_else: else_block,
        },
        span,
    ))
}

/// Parse a `while` statement.
///
/// # Panics
/// Panics if the next token isn't `while`.
pub fn parse_while_stmt(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    let while_tok = par.assert(TokenKind::While);

    let test = parse_expr(par)?;
    par.enter_block(while_tok.span + test.span, "`while` statement")?;
    let body = parse_block_stmts(par)?;
    let span = while_tok.span + test.span + body.last();

    Ok(Node::new(FuncStmt::While { test, body }, span))
}

/// Parse a `for` statement.
///
/// # Panics
/// Panics if the next token isn't `for`.
pub fn parse_for_stmt(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    let for_tok = par.assert(TokenKind::For);

    let target = par
        .expect(TokenKind::Name, "failed to parse `for` statement")?
        .into();
    par.expect(TokenKind::In, "failed to parse `for` statement")?;
    let iter = parse_expr(par)?;
    par.enter_block(for_tok.span + iter.span, "`for` statement")?;
    let body = parse_block_stmts(par)?;
    let span = for_tok.span + iter.span + body.last();

    Ok(Node::new(FuncStmt::For { target, iter, body }, span))
}

/// Parse a `return` statement.
///
/// # Panics
/// Panics if the next token isn't `return`.
pub fn parse_return_stmt(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    let ret = par.assert(TokenKind::Return);
    let value = match par.peek() {
        None | Some(TokenKind::Newline) => None,
        Some(_) => Some(parse_expr(par)?),
    };
    par.expect_newline("return statement")?;
    let span = ret.span + value.as_ref();
    Ok(Node::new(FuncStmt::Return { value }, span))
}

/// Parse an `assert` statement.
///
/// # Panics
/// Panics if the next token isn't `assert`.
pub fn parse_assert_stmt(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    let assert_tok = par.assert(TokenKind::Assert);
    let test = parse_expr(par)?;
    let msg = match par.peek() {
        None | Some(TokenKind::Newline) => None,
        Some(TokenKind::Comma) => {
            par.next()?;
            Some(parse_expr(par)?)
        }
        Some(_) => {
            let tok = par.next()?;
            par.unexpected_token_error(tok.span, "failed to parse `assert` statement", vec![]);
            return Err(ParseFailed);
        }
    };
    par.expect_newline("assert statement")?;
    let span = assert_tok.span + test.span + msg.as_ref();
    Ok(Node::new(FuncStmt::Assert { test, msg }, span))
}

/// Parse an `emit` statement
///
/// # Panics
/// Panics if the next token isn't `emit`
pub fn parse_emit_statement(par: &mut Parser) -> ParseResult<Node<FuncStmt>> {
    let emit_tok = par.assert(TokenKind::Emit);
    let event_name = par.expect(TokenKind::Name, "failed to parse emit statement")?;

    if let Some(TokenKind::ParenOpen) = par.peek() {
        let args = parse_call_args(par)?;
        par.expect_newline("emit statement")?;
        let span = emit_tok.span + args.span;
        let name = Node::new(event_name.text.to_string(), event_name.span);
        return Ok(Node::new(FuncStmt::Emit { name, args }, span));
    } else {
        par.expect(
            TokenKind::ParenOpen,
            "failed to parse event invocation parameter list",
        )?;
    }

    Err(ParseFailed)
}
