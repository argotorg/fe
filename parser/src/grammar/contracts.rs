use super::functions::parse_fn_def;
use super::types::{parse_event_def, parse_field, parse_opt_qualifier};

use crate::ast::ModuleStmt;
use crate::node::Node;
use crate::{ParseFailed, ParseResult, Parser, TokenKind};

// Rule: all "statement" level parse functions consume their trailing
// newline(s), either directly or via a function they call.
// This is required to parse an `if` block, because we need to peek past the
// trailing newlines to check whether it's followed by an `else` block, and is
// done for all statements for consistency.

/// Parse a contract definition.
/// # Panics
/// Panics if the next token isn't `contract`.
pub fn parse_contract_def(par: &mut Parser) -> ParseResult<Node<ModuleStmt>> {
    use TokenKind::*;
    let contract_tok = par.assert(Contract);

    // contract Foo:
    //   x: map<address, u256>
    //   pub y: u8
    //   const z: u256 = 10
    //
    //   event Sent:
    //     idx sender: address
    //     val: u256
    //
    //   pub def foo() -> address:
    //     return abc
    //

    let contract_name = par.expect_with_notes(
        Name,
        "failed to parse contract definition",
        || vec!["Note: `contract` must be followed by a name, which must start with a letter and contain only letters, numbers, or underscores".into()],
    )?;

    let header_span = contract_tok.span + contract_name.span;
    par.enter_block(header_span, "contract definition")?;

    let mut fields = vec![];
    let mut defs = vec![];

    loop {
        let mut pub_qual = parse_opt_qualifier(par, TokenKind::Pub);
        let const_qual = parse_opt_qualifier(par, TokenKind::Const);
        if pub_qual.is_none() && const_qual.is_some() && par.peek() == Some(Pub) {
            pub_qual = parse_opt_qualifier(par, TokenKind::Pub);
            par.error(
                pub_qual.unwrap() + const_qual,
                "`const pub` should be written `pub const`",
            );
        }

        match par.peek() {
            Some(TokenKind::Name) => {
                let field = parse_field(par, pub_qual, const_qual)?;
                if !defs.is_empty() {
                    par.error(field.span, "contract field definitions must come before any function or event definitions");
                }
                fields.push(field);
            }
            Some(TokenKind::Def) => {
                if let Some(span) = const_qual {
                    par.error(
                        span,
                        "`const` qualifier can't be used with function definitions",
                    );
                }
                defs.push(parse_fn_def(par, pub_qual)?);
            }
            Some(TokenKind::Event) => {
                if let Some(span) = pub_qual {
                    par.error(span, "`pub` qualifier can't be used with event definitions");
                }
                if let Some(span) = const_qual {
                    par.error(
                        span,
                        "`const` qualifier can't be used with event definitions",
                    );
                }
                defs.push(parse_event_def(par)?);
            }
            Some(TokenKind::Dedent) => {
                par.next()?;
                break;
            }
            None => break,
            Some(_) => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    tok.span,
                    "failed to parse contract definition body",
                    vec![],
                );
                return Err(ParseFailed);
            }
        };
    }

    if fields.is_empty() && defs.is_empty() {
        par.error(header_span, "empty contract definition");
    }
    let span = header_span + fields.last() + defs.last();
    Ok(Node::new(
        ModuleStmt::ContractDef {
            name: Node::new(contract_name.text.to_string(), contract_name.span),
            fields,
            body: defs,
        },
        span,
    ))
}
