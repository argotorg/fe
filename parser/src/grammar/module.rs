//! Parsing function for statements.
use super::contracts::parse_contract_def;
use super::types::{
    parse_struct_def,
    parse_type_def,
};
use crate::ast::{
    FromImportPath,
    Module,
    ModuleStmt,
    SimpleImportName,
};
use crate::node::{
    Node,
    Span,
};
use crate::{
    ParseFailed,
    ParseResult,
    Parser,
    TokenKind,
};

pub fn parse_module(par: &mut Parser) -> ParseResult<Node<Module>> {
    let mut body = vec![];
    loop {
        match par.peek() {
            Some(TokenKind::Newline) => par.expect_newline("module")?,
            Some(TokenKind::Dedent) => {
                par.next()?;
                break;
            }
            None => break,
            Some(_) => {
                let stmt = parse_module_stmt(par)?;
                body.push(stmt);
            }
        }
    }
    let span = if body.is_empty() {
        Span::new(0, 0)
    } else {
        body.first().unwrap().span + body.last()
    };

    Ok(Node::new(Module { body }, span))
}

pub fn parse_module_stmt(par: &mut Parser) -> ParseResult<Node<ModuleStmt>> {
    match par.peek().unwrap() {
        TokenKind::Import => parse_simple_import(par),
        TokenKind::Name if par.peeked_text() == "from" => parse_from_import(par),
        TokenKind::Contract => parse_contract_def(par),
        TokenKind::Struct => parse_struct_def(par),
        TokenKind::Type => parse_type_def(par),
        TokenKind::Event => todo!("module-level event def"),
        _ => {
            let tok = par.next().unwrap();
            par.unexpected_token_error(
                tok.span,
                "failed to parse module",
                vec!["Note: expected import, contract, struct, type, or event".into()],
            );
            par.error(
                tok.span,
                format!("Unexpected token when parsing module: `{:?}`", tok.kind),
            );
            Err(ParseFailed)
        }
    }
}

pub fn parse_simple_import(par: &mut Parser) -> ParseResult<Node<ModuleStmt>> {
    let import_tok = par.assert(TokenKind::Import);

    // TODO: only handles `import foo, bar as baz`

    let mut names = vec![];
    loop {
        let name =
            par.expect_with_notes(TokenKind::Name, "failed to parse import statement", || {
                vec![
                    "Note: `import` must be followed by a module name or path".into(),
                    "Example: `import mymodule".into(),
                ]
            })?;

        let alias = if par.peek() == Some(TokenKind::As) {
            par.next()?;
            let tok = par.expect(TokenKind::Name, "failed to parse import statement")?;
            Some(tok.into())
        } else {
            None
        };

        let span = name.span + alias.as_ref();
        names.push(Node::new(
            SimpleImportName {
                path: vec![Node::new(name.text.to_string(), name.span)],
                alias,
            },
            span,
        ));

        match par.peek() {
            Some(TokenKind::Comma) => {
                par.next()?;
                continue;
            }
            Some(TokenKind::Newline) | None => break,
            Some(_) => {
                let tok = par.next()?;
                par.unexpected_token_error(tok.span, "failed to parse `import` statement", vec![]);
                return Err(ParseFailed);
            }
        }
    }

    let span = import_tok.span + names.last();
    Ok(Node::new(ModuleStmt::SimpleImport { names }, span))
}

pub fn parse_import_path(_par: &mut Parser) -> ParseResult<FromImportPath> {
    todo!("parse import path (not supported in rest of compiler yet)")
}

pub fn parse_from_import(par: &mut Parser) -> ParseResult<Node<ModuleStmt>> {
    let tok = par.assert(TokenKind::Name);
    assert_eq!(tok.text, "from");
    todo!("parse from .. import (not supported in rest of compiler yet)")
}
