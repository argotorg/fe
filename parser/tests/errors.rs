use fe_common::diagnostics::diagnostics_string;
use fe_parser::grammar::{contracts, expressions, functions, module, types};

use fe_parser::{ParseResult, Parser};
use insta::assert_snapshot;

pub fn err_string<F, T>(mut parse_fn: F, should_fail: bool, src: &str) -> String
where
    F: FnMut(&mut Parser) -> ParseResult<T>,
    T: std::fmt::Debug,
{
    let mut files = fe_common::files::FileStore::new();
    let id = files.add_file("[test snippet]", src);
    let mut parser = Parser::new(src, id);

    if should_fail != parse_fn(&mut parser).is_err() {
        panic!(
            "expected parsing to {}fail",
            if should_fail { "" } else { "not " }
        );
    }
    diagnostics_string(&parser.diagnostics, &files)
}

macro_rules! test_parse_err {
    ($name:ident, $parse_fn:expr, $should_fail:expr, $string:expr) => {
        #[test]
        fn $name() {
            let err = err_string($parse_fn, $should_fail, $string);
            assert_snapshot!(err);
        }
    };
}

// These tests use the insta crate. insta will automatically generate the
// snapshot file on the first run.

test_parse_err! { contract_bad_name, contracts::parse_contract_def, true, "contract 1X:\n x: u8" }
test_parse_err! { contract_empty_body, module::parse_module, true, "contract X:\n \n \ncontract Y:\n x: u8" }
test_parse_err! { contract_field_after_def, module::parse_module, false, r#"
contract C:
  def f():
    pass
  x: u8
"#
}

test_parse_err! { contract_pub_event, contracts::parse_contract_def, false, "contract C:\n pub event E:\n  x: u8" }
test_parse_err! { contract_const_pub, contracts::parse_contract_def, false, "contract C:\n const pub x: u8" }
test_parse_err! { contract_const_fn, contracts::parse_contract_def, false, "contract C:\n const def f():\n  pass" }
test_parse_err! { emit_no_args, functions::parse_stmt, true, "emit x" }
test_parse_err! { emit_expr, functions::parse_stmt, true, "emit x + 1" }
test_parse_err! { emit_bad_call, functions::parse_stmt, true, "emit MyEvent(1)()" }
test_parse_err! { expr_bad_prefix, expressions::parse_expr, true, "*x + 1" }
test_parse_err! { for_no_in, functions::parse_stmt, true, "for x:\n pass" }
test_parse_err! { fn_no_args, |par| functions::parse_fn_def(par, None), false, "def f:\n  return 5" }
test_parse_err! { if_no_body, functions::parse_stmt, true, "if x:\nelse:\n x" }
test_parse_err! { import_bad_name, module::parse_simple_import, true, "import x as 123" }
test_parse_err! { module_bad_stmt, module::parse_module, true, "if x:\n y" }
test_parse_err! { module_nonsense, module::parse_module, true, "))" }
test_parse_err! { string_invalid_escape, expressions::parse_expr, false, r#""a string \c ""# }
test_parse_err! { struct_bad_field_name, types::parse_struct_def, true, "struct f:\n pub def" }
