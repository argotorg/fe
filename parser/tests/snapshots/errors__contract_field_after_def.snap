---
source: parser/tests/errors.rs
expression: err

---
error: contract field definitions must come before any function or event definitions
  ┌─ [test snippet]:5:3
  │
5 │   x: u8
  │   ^^^^^


